omnibus/files/private-chef-cookbooks/private-chef/libraries/preflight_checks.rb

```diff
@@ -47,6 +58,42 @@ def backend?
     EnterpriseChef::Helpers.backend? faux_node
   end
 
+  # Postgresql connectivity and validation functions.
+  def named_db_exists?(connection, name)
+    connection.exec("SELECT count(*) AS result FROM pg_database WHERE datname='#{name}'")[0]['result'] == '1'
+  end
+
+  def named_role_exists?(connection, username)
+    # If a record exists, the role exists:
+    connection.exec("select usesuper from pg_catalog.pg_user where usename = '#{username}'").ntuples > 0
+  end
+
+  def connect_as(type, db_name = 'template1')
+    require "pg" # Make the PG constants available to the caller
+    port = cs_pg_attr.has_key?('port') ? cs_pg_attr['port'] : node_pg_attr['port']
+    host = cs_pg_attr['vip']
+    if type == :invalid_user
+      user = 'chef_server_conn_test'
+      password = 'invalid'
+    else
+      user = cs_pg_attr['db_superuser']
+      password = cs_pg_attr['db_superuser_password']
+    end
+    # We just want this to throw an exception or not - caller knows what to do with it.
+    EcPostgres.with_connection(node, db_name, { 'db_superuser' => user,
+                                                'db_superuser_password' => password,
+                                                'vip' => host,
+                                                'port' => port,
+                                                # By default, pass exceptions up without retrying
+                                                # We have a few scenarios in which we
+                                                # expect errors, and don't want to delay the bootstrap
+                                                # for unnecessary retries.
+                                                'retries' => 0 }) do |conn|
+       if block_given?
+         yield(conn)
+       end
+    end
```

Comment:
connect_as is shared between them.  The other two were done preemptively - when it's time to refine and finalize the bootstrap validator, they will also need to be there. 

I can actually use one of them to get rid of a nested begin/rescue/end block, now that I think about it. I'll make that change, and move the roles checking function back into pg validator until we need it elsewhere.